File: transport_test.go
package yaylib

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"strings"
	"sync"
	"sync/atomic"
	"testing"
	"time"
)

// requestCapturer collects the headers and paths of every incoming
// request, while letting each test plug in a custom handler for the
// response side.
type requestCapturer struct {
	mu       sync.Mutex
	requests []capturedRequest
	handler  func(w http.ResponseWriter, r *http.Request)
}

type capturedRequest struct {
	method  string
	path    string
	headers http.Header
	body    []byte
}

func (c *requestCapturer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	body, _ := io.ReadAll(r.Body)
	r.Body.Close()
	c.mu.Lock()
	c.requests = append(c.requests, capturedRequest{
		method: r.Method, path: r.URL.Path, headers: r.Header.Clone(), body: body,
	})
	c.mu.Unlock()
	c.handler(w, r)
}

func (c *requestCapturer) snapshot() []capturedRequest {
	c.mu.Lock()
	defer c.mu.Unlock()
	out := make([]capturedRequest, len(c.requests))
	copy(out, c.requests)
	return out
}

// PORTING:S6
func TestTransport_InjectsRequiredHeaders(t *testing.T) {
	cap := &requestCapturer{handler: func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{}`))
	}}
	srv := httptest.NewServer(cap)
	defer srv.Close()

	c := NewClient(WithBaseURL(srv.URL))
	c.SetTokens("ACC", "REF")

	req, _ := http.NewRequestWithContext(context.Background(), http.MethodGet, srv.URL+"/v1/anywhere", nil)
	resp, err := c.httpClient.Do(req)
	if err != nil {
		t.Fatalf("Do: %v", err)
	}
	resp.Body.Close()

	got := cap.snapshot()[0].headers
	for _, h := range []string{
		"User-Agent", "X-App-Version", "X-Device-Info", "X-Device-UUID",
		"X-Connection-Type", "X-Connection-Speed", "Accept-Language",
		"X-Timestamp",
	} {
		if got.Get(h) == "" {
			t.Errorf("missing required header %s", h)
		}
	}
	if got.Get("Authorization") != "Bearer ACC" {
		t.Errorf("Authorization = %q, want %q", got.Get("Authorization"), "Bearer ACC")
	}
}

// PORTING:S6
func TestTransport_OAuthEndpointUsesBasicAuth(t *testing.T) {
	cap := &requestCapturer{handler: func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{}`))
	}}
	srv := httptest.NewServer(cap)
	defer srv.Close()

	c := NewClient(WithBaseURL(srv.URL))
	c.SetTokens("ACC", "REF")

	req, _ := http.NewRequestWithContext(context.Background(), http.MethodPost, srv.URL+"/api/v1/oauth/token", strings.NewReader(""))
	resp, err := c.httpClient.Do(req)
	if err != nil {
		t.Fatalf("Do: %v", err)
	}
	resp.Body.Close()

	got := cap.snapshot()[0].headers.Get("Authorization")
	if !strings.HasPrefix(got, "Basic ") {
		t.Errorf("Authorization on /api/v1/oauth/token = %q, want Basic prefix", got)
	}
}

// expired-token: protected requests 401 (error_code -3) until the
// token endpoint is hit (a simulated refresh), then the happy path.
// This is the 401 → refresh → retry chain. Behavioral assertion:
// final success and the seeded tokens rotated to the server-issued
// ones (the Bearer-header progression is a Go-internal mechanic, and
// §15 says translate the scenario, not the mechanics).
//
// PORTING:S7
func TestTransport_401TriggersRefreshAndRetries(t *testing.T) {
	c := mockClient(t, "expired-token", WithRetryPolicy(RetryPolicy{}))
	c.SetTokens("STALE", "REF")

	resp, err := mockGet(t, c)
	if err != nil {
		t.Fatalf("Get: %v", err)
	}
	if resp.StatusCode != http.StatusOK {
		t.Errorf("final status = %d, want 200 (refresh + retry succeeded)", resp.StatusCode)
	}
	tok := c.Tokens()
	if tok.Access == "STALE" || tok.Access == "" {
		t.Errorf("access token = %q, want it rotated to the server-issued value", tok.Access)
	}
	if tok.Refresh == "REF" || tok.Refresh == "" {
		t.Errorf("refresh token = %q, want it rotated to the server-issued value", tok.Refresh)
	}
}

// fail-401-times-2: the data request 401s and the refresh call (same
// session+scenario counter) also 401s, so the refresh fails and the
// original 401 surfaces with its body intact (README composition note).
//
// PORTING:S8
func TestTransport_401RefreshFailureSurfaces401(t *testing.T) {
	c := mockClient(t, "fail-401-times-2", WithRetryPolicy(RetryPolicy{}))
	c.SetTokens("STALE", "REF")

	resp, err := mockGetRaw(t, c)
	if err != nil {
		t.Fatalf("Get: %v", err)
	}
	body, _ := io.ReadAll(resp.Body)
	resp.Body.Close()

	if resp.StatusCode != http.StatusUnauthorized {
		t.Errorf("status = %d, want 401 (refresh failed, original 401 surfaced)", resp.StatusCode)
	}
	if len(body) == 0 {
		t.Error("body empty; want the original 401 error body preserved")
	}
}

// PORTING:S9
func TestTransport_401RefreshSucceedsButRetryNetworkErrorSurfacesError(t *testing.T) {
	// Server: oauth/token returns fresh tokens. Data path returns 401
	// once, then closes the connection on the retry to simulate a
	// post-refresh network failure.
	var dataAttempt int32
	srv := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if strings.Contains(r.URL.Path, "/api/v1/oauth/token") {
			w.Header().Set("Content-Type", "application/json")
			_ = json.NewEncoder(w).Encode(map[string]any{
				"access_token":  "NEW_ACC",
				"refresh_token": "NEW_REF",
			})
			return
		}
		n := atomic.AddInt32(&dataAttempt, 1)
		if n == 1 {
			w.WriteHeader(http.StatusUnauthorized)
			return
		}
		// Force a connection-level failure on the retry by hijacking
		// and abruptly closing the underlying TCP socket.
		hj, ok := w.(http.Hijacker)
		if !ok {
			t.Fatal("server doesn't support hijack")
		}
		conn, _, err := hj.Hijack()
		if err != nil {
			t.Fatalf("hijack: %v", err)
		}
		_ = conn.Close()
	}))
	srv.Start()
	defer srv.Close()

	c := NewClient(WithBaseURL(srv.URL), WithRetryPolicy(RetryPolicy{}))
	c.SetTokens("STALE", "REF")

	req, _ := http.NewRequestWithContext(context.Background(), http.MethodGet, srv.URL+"/v1/anywhere", nil)
	resp, err := c.httpClient.Do(req)
	if err == nil {
		// httpClient.Do may surface as resp + 0 bytes if the connection
		// was closed mid-response, but our handle401 path returns
		// (nil, err) so we expect err non-nil here.
		resp.Body.Close()
		t.Fatal("expected error from broken retry, got nil")
	}
	if c.Tokens().Access != "NEW_ACC" {
		t.Errorf("tokens should still be rotated, got access=%q", c.Tokens().Access)
	}
}

// PORTING:S10
func TestTransport_LazyFetchClientIPPopulatesHeader(t *testing.T) {
	const fetchedIP = "203.0.113.7"
	cap := &requestCapturer{handler: func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		if strings.HasSuffix(r.URL.Path, "/v2/users/timestamp") {
			_ = json.NewEncoder(w).Encode(map[string]any{
				"time":       int64(0),
				"ip_address": fetchedIP,
			})
			return
		}
		_, _ = w.Write([]byte(`{}`))
	}}
	srv := httptest.NewServer(cap)
	defer srv.Close()

	c := NewClient(WithBaseURL(srv.URL))

	// First request kicks off the lazy IP fetch in a background
	// goroutine; the request itself goes out without X-Client-IP.
	req1, _ := http.NewRequestWithContext(context.Background(), http.MethodGet, srv.URL+"/v1/first", nil)
	resp1, err := c.httpClient.Do(req1)
	if err != nil {
		t.Fatalf("first Do: %v", err)
	}
	resp1.Body.Close()

	deadline := time.Now().Add(2 * time.Second)
	for time.Now().Before(deadline) {
		if c.getClientIP() == fetchedIP {
			break
		}
		time.Sleep(20 * time.Millisecond)
	}
	if c.getClientIP() != fetchedIP {
		t.Fatalf("ClientIP not populated after lazy fetch, got %q", c.getClientIP())
	}

	req2, _ := http.NewRequestWithContext(context.Background(), http.MethodGet, srv.URL+"/v1/second", nil)
	resp2, err := c.httpClient.Do(req2)
	if err != nil {
		t.Fatalf("second Do: %v", err)
	}
	resp2.Body.Close()

	requests := cap.snapshot()
	var firstIP, secondIP string
	for _, req := range requests {
		switch req.path {
		case "/v1/first":
			firstIP = req.headers.Get("X-Client-IP")
		case "/v1/second":
			secondIP = req.headers.Get("X-Client-IP")
		}
	}
	if firstIP != "" {
		t.Errorf("first request X-Client-IP = %q, want empty (fetch is async)", firstIP)
	}
	if secondIP != fetchedIP {
		t.Errorf("second request X-Client-IP = %q, want %q", secondIP, fetchedIP)
	}
}

File: packages/typescript/tests/transport.test.ts
// Transport behavior parity (PORTING.md §15 #6a / #6b / #10 / #9).
//
// These exercise the request-shaping the SDK guarantees on every
// outbound call against an in-process HTTP server stub:
//
//   #6a  the required headers (User-Agent / X-App-Version / X-Device-*
//        / X-Connection-* / Accept-Language / X-Timestamp) ride every
//        request.
//   #6b  the OAuth token endpoint gets `Authorization: Basic`, while a
//        normal endpoint gets `Authorization: Bearer` once tokens are
//        set.
//   #10  the caller's outbound IP is learned lazily: the first request
//        carries no X-Client-IP, and a later request does once the SDK
//        has observed it from a server response.
//   #9   a 401 that refreshes successfully but whose retried original
//        then hits a transport error surfaces that transport error —
//        never a stale 401.

import { createServer, type Server } from "node:http";
import type { AddressInfo } from "node:net";

import { Client } from "../src/client";

let failed = 0;
function assert(name: string, cond: boolean, detail = ""): void {
  if (cond) console.log(`PASS ${name}`);
  else {
    failed++;
    console.log(`FAIL ${name}${detail ? `\n  ${detail}` : ""}`);
  }
}

interface Captured {
  method: string;
  path: string;
  headers: Record<string, string>;
}

type Reply = { status: number; body: string; headers?: Record<string, string> };
type HandlerFn = (req: Captured, body: string) => Reply | "destroy";

// withServer runs an in-process server that records every request's
// method / path / headers, then defers the response to the handler. A
// handler returning "destroy" kills the socket so the client sees a
// transport-level failure (the TS analogue of the Go test hijacking and
// closing the TCP connection on the post-refresh retry).
async function withServer(
  handler: HandlerFn,
  fn: (baseURL: string, captured: Captured[]) => Promise<void>,
): Promise<void> {
  const captured: Captured[] = [];
  const server: Server = createServer((req, res) => {
    const chunks: Buffer[] = [];
    req.on("data", (c) => chunks.push(c));
    req.on("end", () => {
      const headers: Record<string, string> = {};
      for (const [k, v] of Object.entries(req.headers)) {
        if (typeof v === "string") headers[k.toLowerCase()] = v;
        else if (Array.isArray(v)) headers[k.toLowerCase()] = v.join(", ");
      }
      const rec: Captured = {
        method: req.method ?? "GET",
        path: (req.url ?? "").split("?")[0],
        headers,
      };
      captured.push(rec);
      const out = handler(rec, Buffer.concat(chunks).toString("utf8"));
      if (out === "destroy") {
        req.socket.destroy();
        return;
      }
      res.writeHead(out.status, {
        "Content-Type": "application/json",
        ...(out.headers ?? {}),
      });
      res.end(out.body);
    });
  });
  await new Promise<void>((r) => server.listen(0, r));
  const baseURL = `http://127.0.0.1:${(server.address() as AddressInfo).port}`;
  try {
    await fn(baseURL, captured);
  } finally {
    server.closeAllConnections?.();
    await new Promise<void>((r) => server.close(() => r()));
  }
}

const NO_RETRY = { maxAttempts: 0, baseDelay: 1, maxDelay: 1, retryOnPOST: false };

function get(headers: Record<string, string>, key: string): string {
  return headers[key.toLowerCase()] ?? "";
}

// #6a — every request carries the required headers; with tokens set the
// Authorization scheme is Bearer on a normal endpoint.
async function requiredHeadersInjected(): Promise<void> {
  await withServer(
    () => ({ status: 200, body: "{}" }),
    async (baseURL, captured) => {
      const c = new Client({ baseURL, retryPolicy: NO_RETRY });
      c.setTokens("ACC", "REF");
      await c.bucketsAPI.getBucketPresignedUrls({ fileNames: [] });

      const h = captured[0].headers;
      for (const name of [
        "User-Agent",
        "X-App-Version",
        "X-Device-Info",
        "X-Device-UUID",
        "X-Connection-Type",
        "X-Connection-Speed",
        "Accept-Language",
        "X-Timestamp",
      ]) {
        assert(`required header present: ${name}`, get(h, name) !== "", `${name} empty`);
      }
      assert(
        "Authorization is Bearer on a normal endpoint",
        get(h, "Authorization") === "Bearer ACC",
        get(h, "Authorization"),
      );
    },
  );
}

// #6b — the OAuth token endpoint gets Basic; a normal endpoint gets
// Bearer. Asserted on the same client so the scheme selection is the
// transport's, not the caller's.
async function oauthEndpointUsesBasicAuth(): Promise<void> {
  await withServer(
    (req) => {
      if (req.path === "/api/v1/oauth/token") {
        return {
          status: 200,
          body: JSON.stringify({ access_token: "A", refresh_token: "R" }),
        };
      }
      return { status: 200, body: "{}" };
    },
    async (baseURL, captured) => {
      const c = new Client({ baseURL, retryPolicy: NO_RETRY });
      c.setTokens("ACC", "REF");

      await c.authAPI.oauthToken({ grantType: "refresh_token", refreshToken: "R" });
      await c.bucketsAPI.getBucketPresignedUrls({ fileNames: [] });

      const oauthReq = captured.find((r) => r.path === "/api/v1/oauth/token");
      const normalReq = captured.find((r) => r.path === "/v1/buckets/presigned_urls");
      assert("oauth request captured", oauthReq !== undefined);
      assert("normal request captured", normalReq !== undefined);

      const oauthAuth = get(oauthReq!.headers, "Authorization");
      const expectedBasic = `Basic ${Buffer.from(c.apiKey).toString("base64")}`;
      assert(
        "oauth endpoint Authorization is Basic <base64(apiKey)>",
        oauthAuth === expectedBasic,
        `got=${oauthAuth} want=${expectedBasic}`,
      );
      assert(
        "normal endpoint Authorization is Bearer after setTokens",
        get(normalReq!.headers, "Authorization") === "Bearer ACC",
        get(normalReq!.headers, "Authorization"),
      );
    },
  );
}

// #10 — lazy X-Client-IP. The first request carries none; the SDK learns
// the IP from the timestamp lookup response; a later request carries it.
async function lazyClientIPPopulatesHeader(): Promise<void> {
  const FETCHED_IP = "203.0.113.7";
  await withServer(
    (req) => {
      if (req.path.endsWith("/v2/users/timestamp")) {
        return {
          status: 200,
          body: JSON.stringify({ time: 0, ip_address: FETCHED_IP }),
        };
      }
      return { status: 200, body: "{}" };
    },
    async (baseURL, captured) => {
      const c = new Client({ baseURL, retryPolicy: NO_RETRY });

      // First request kicks off the background IP lookup; it goes out
      // without X-Client-IP.
      await c.bucketsAPI.getBucketPresignedUrls({ fileNames: [] });

      // Wait until the SDK has observed the IP (or time out).
      const deadline = Date.now() + 2000;
      while (Date.now() < deadline && c.clientIP !== FETCHED_IP) {
        await new Promise((r) => setTimeout(r, 20));
      }
      assert(
        "client IP learned from lookup response",
        c.clientIP === FETCHED_IP,
        `clientIP=${c.clientIP}`,
      );

      // Second request now carries the learned X-Client-IP.
      await c.bucketsAPI.getBucketPresignedUrls({ fileNames: [] });

      const bucketReqs = captured.filter(
        (r) => r.path === "/v1/buckets/presigned_urls",
      );
      const first = bucketReqs[0];
      const second = bucketReqs[bucketReqs.length - 1];
      assert(
        "first request carries no X-Client-IP (lookup is async)",
        get(first.headers, "X-Client-IP") === "",
        get(first.headers, "X-Client-IP"),
      );
      assert(
        "second request carries the learned X-Client-IP",
        get(second.headers, "X-Client-IP") === FETCHED_IP,
        get(second.headers, "X-Client-IP"),
      );
    },
  );
}

// #9 — 401 → refresh succeeds → retried original hits a transport error.
// The surfaced error must be the retry/network error, not a stale 401.
async function refreshOkButRetryNetworkErrorSurfaces(): Promise<void> {
  let dataAttempt = 0;
  await withServer(
    (req): Reply | "destroy" => {
      if (req.path === "/api/v1/oauth/token") {
        return {
          status: 200,
          body: JSON.stringify({
            access_token: "NEW_ACC",
            refresh_token: "NEW_REF",
          }),
        };
      }
      dataAttempt++;
      if (dataAttempt === 1) {
        return { status: 401, body: '{"error_code":-3,"message":"expired"}' };
      }
      // Post-refresh retry: kill the socket so fetch rejects with a
      // transport error.
      return "destroy";
    },
    async (baseURL) => {
      const c = new Client({ baseURL, retryPolicy: NO_RETRY });
      c.setTokens("STALE", "REF");

      let threw = false;
      let status401 = false;
      try {
        await c.bucketsAPI.getBucketPresignedUrls({ fileNames: [] });
      } catch (err) {
        threw = true;
        // A surfaced ResponseError/APIError carrying a 401 would mean
        // the stale 401 leaked — assert it did not.
        const anyErr = err as { response?: { status?: number } };
        status401 = anyErr.response?.status === 401;
      }
      assert("broken retry surfaces an error", threw);
      assert(
        "surfaced error is the transport failure, not a stale 401",
        threw && !status401,
      );
      assert(
        "tokens stayed rotated after the successful refresh",
        c.tokens().access === "NEW_ACC",
        `access=${c.tokens().access}`,
      );
    },
  );
}

(async () => {
  await requiredHeadersInjected();
  await oauthEndpointUsesBasicAuth();
  await lazyClientIPPopulatesHeader();
  await refreshOkButRetryNetworkErrorSurfaces();
  process.exit(failed === 0 ? 0 : 1);
})().catch((e) => {
  console.error("test crashed:", e);
  process.exit(2);
});
